google/cloud/storage/internal/async/rewriter_connection_tracing.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/async/rewriter_connection_tracing.h"
+#include "google/cloud/storage/async/rewriter_connection.h"
+#include "google/cloud/internal/opentelemetry.h"
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace {
+
+using google::cloud::internal::EndSpan;
+
+class AsyncRewriterTracingConnection
+    : public storage_experimental::AsyncRewriterConnection {
+ public:
+  AsyncRewriterTracingConnection(
+      std::shared_ptr<storage_experimental::AsyncRewriterConnection> impl,
+      opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span)
+      : impl_(std::move(impl)), span_(std::move(span)) {}
+  ~AsyncRewriterTracingConnection() override { EndSpan(*span_, Status{}); }
+
+  future<StatusOr<storage_experimental::RewriteObjectResponse>> Iterate()
+      override {
+    internal::OTelScope scope(span_);
+    return impl_->Iterate().then([span = span_](auto f) {
+      auto r = f.get();
```

Comment:
I don't think we want the span active in continuations of the returned future. `DetachOTelContext(...)` is how we do this. e.g.: 
https://github.com/googleapis/google-cloud-cpp/blob/5a1fe557ae3c37e0611322099dde29e4bfd0d730/google/cloud/storage/internal/async/connection_tracing.cc#L72-L72

The order of `DetachOTelContext()` and `EndSpan()` doesn't matter, so something like:
```cc
    return impl_->Iterate().then([span = span_, oc = opentelemetry::context::RuntimeContext::GetCurrent()](auto f) {
      auto r = f.get();
      internal::DetachOTelContext(oc);
      // etc...
```

Reply:
Doen